File: 프로그래머스/3/118669. 등산코스 정하기/등산코스 정하기.cpp
#include <string>
#include <vector>
#include <queue>
#include <climits>
#include <algorithm>

using namespace std;

vector<int> solution(int n, vector<vector<int>> paths, vector<int> gates, vector<int> summits) {    
    vector<vector<pair<int, int>>> edgeList(n+1);
    for (auto p : paths) {
        edgeList[p[0]].push_back({p[1], p[2]});
        edgeList[p[1]].push_back({p[0], p[2]});
    }
    
    vector<bool> isSummit(n+1, false);
    for (int s : summits) isSummit[s] = true;
    
    priority_queue<pair<int, int>, vector<pair<int, int>>, greater<>> pq;
    vector<int> intensity(n+1, INT_MAX);
    
    for (int gate : gates) {
        intensity[gate] = 0;
        pq.push({0, gate});
    }
    
    while(!pq.empty()) {
        auto [curIntensity, cur] = pq.top(); pq.pop();
        
        if (curIntensity > intensity[cur]) continue;
        if (isSummit[cur]) continue;
        
        for (auto [next, w] : edgeList[cur]) {
            int maxIntensity = max(w, curIntensity);
            if (intensity[next] > maxIntensity) {
                intensity[next] = maxIntensity;
                pq.push({intensity[next], next});
            }
        }
    }
    
    sort(summits.begin(), summits.end());
    
    int bestIntensity = INT_MAX;
    int bestNode = 0;
    for (int s : summits) {
        if (intensity[s] < bestIntensity) {
            bestIntensity = intensity[s];
            bestNode = s;
        }
    }
    
    return {bestNode, bestIntensity};
}

File: 프로그래머스/3/118669. 등산코스 정하기/등산코스 정하기.java
import java.util.*;

class Solution {
    List<int[]>[] graph;
    boolean[] isSummit;
    int[] dist;
    
    public int[] solution(int n, int[][] paths, int[] gates, int[] summits) {
        graph = new ArrayList[n + 1];
        isSummit = new boolean[n + 1];
        
        for (int i = 1; i <= n; i++) {
            graph[i] = new ArrayList<>();
        }
        
        for (int[] path : paths) {
            int a = path[0];
            int b = path[1];
            int weight = path[2];
            
            graph[a].add(new int[]{b, weight});
            graph[b].add(new int[]{a, weight});
        }
        
        for (int summit : summits) {
            isSummit[summit] = true;
        }
        
        dijkstra(n, gates);
        
        Arrays.sort(summits);
        
        int selectedSummit = 0;
        int minIntensity = Integer.MAX_VALUE;
        
        for (int summit : summits) {
            if (dist[summit] < minIntensity) {
                minIntensity = dist[summit];
                selectedSummit = summit;
            }
        }
        
        return new int[]{selectedSummit, minIntensity};
    }
    
    private void dijkstra(int n, int[] gates) {
        PriorityQueue<int[]> pq = new PriorityQueue<>((a, b) -> Integer.compare(a[1], b[1]));
        
        dist = new int[n + 1];
        Arrays.fill(dist, Integer.MAX_VALUE);
        
        for (int gate : gates) {
            dist[gate] = 0;
            pq.offer(new int[]{gate, 0});
        }

        while(!pq.isEmpty()) {
            int[] cur = pq.poll();
            
            int node = cur[0];
            int intensity = cur[1];
            
            if (dist[node] < intensity) {
                continue;
            }
            
            if (isSummit[node]) {
                continue;
            }
            
            for (int[] next : graph[node]) {
                int nextNode = next[0];
                int edgeWeight = next[1];
                
                int nextIntensity = Math.max(intensity, edgeWeight);
                
                if (dist[nextNode] > nextIntensity) {
                    dist[nextNode] = nextIntensity;
                    pq.offer(new int[]{nextNode, nextIntensity});
                }
            }
        }
    }
}